### PR TITLE
feat: grid based surface material

### PR DESCRIPTION
**Pull request by asalzburger**
This PR superseds #2701 and thus:

* introduces a new Surface material type based on grids
* harmonises the castPosition() method and encapsulates it in a dedicated file (used by the new GridSurfaceMaterial as well)
* moves the GridAxisGenerators from detail into Utilities as commonly used now

This includes the comments of @niermann999 already on the first PR.

### CODE REVIEW

File: Core/include/Acts/Material/GridSurfaceMaterial.hpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    auto index = grid.atPosition(point);
+    return material[index];
+  }
+
+  /// @brief Scale the material (by scaling the thickness)
+  ///
+  /// @param grid the grid (ignored)
+  /// @param scale the amount of the scaling
+  template <typename grid_type>
+  void scale([[maybe_unused]] grid_type& grid, ActsScalar scale) {
```
Comment (niermann999) [inline Core/include/Acts/Material/GridSurfaceMaterial.hpp:74]: I would rather
```suggestion
  void scale(grid_type& /*grid*/, ActsScalar scale) {
```

File: Core/include/Acts/Material/GridSurfaceMaterial.hpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Material/ISurfaceMaterial.hpp"
+#include "Acts/Material/MaterialSlab.hpp"
+#include "Acts/Utilities/BinningType.hpp"
+#include "Acts/Utilities/GridAccessHelpers.hpp"
+#include "Acts/Utilities/VectorHelpers.hpp"
+
+#include <vector>
```
Comment (niermann999) [inline Core/include/Acts/Material/GridSurfaceMaterial.hpp:18]:
```suggestion
#include <ostream>
#include <stdexcept>
#include <vector>
```

File: Tests/UnitTests/Core/Material/CMakeLists.txt [hunk 3]
```diff
@@ -3,6 +3,7 @@ add_unittest(AccumulatedSurfaceMaterial AccumulatedSurfaceMaterialTests.cpp)
 add_unittest(AccumulatedVolumeMaterial AccumulatedVolumeMaterialTests.cpp)
 add_unittest(AverageMaterials AverageMaterialsTests.cpp)
 add_unittest(BinnedSurfaceMaterial BinnedSurfaceMaterialTests.cpp)
+#add_unittest(GridSurfaceMaterial GridSurfaceMaterialTests.cpp)
```
Comment (niermann999) [inline Tests/UnitTests/Core/Material/CMakeLists.txt:6]: Is this supposed to be commented out?

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  return casted;
+}
+
+/// Unroll the local position loop
+///
+/// @param lposition is the local position
+/// @param laccess the local accessors
+/// @param ra is the array to be filled
+///
+/// @note voide function that fill the provided array
```
Comment (niermann999) [inline Core/include/Acts/Utilities/GridAccessHelpers.hpp:63]:
```suggestion
/// @note void function that fills the provided array
```

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+/// Helper method to convert:
+/// 1-D grids:
+/// * [ 0 - n ]
+/// * [ 0 - m ]
+/// 2-D grid:
+/// * [ 0 - n ] x [ 0, m ]
+///
+/// To a local position that can be looked up the accessLocal(...) method
```
Comment (niermann999) [inline Core/include/Acts/Utilities/GridAccessHelpers.hpp:102]:
```suggestion
/// To a local position that can be looked up via the accessLocal(...) method
```

File: Tests/UnitTests/Core/Material/GridSurfaceMaterialTests.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
```
Comment (niermann999) [inline Tests/UnitTests/Core/Material/GridSurfaceMaterialTests.cpp:3]:
```suggestion
// Copyright (C) 2024 CERN for the benefit of the Acts project
```

**Review by niermann999 — COMMENTED** [review]
Just a few very minor comments, otherwise looks good

File: Tests/UnitTests/Core/Material/CMakeLists.txt (hunk 3 above)
Comment (asalzburger) [reply Tests/UnitTests/Core/Material/CMakeLists.txt:6]: Oh, I need to add this.

File: Core/include/Acts/Material/GridSurfaceMaterial.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Material/GridSurfaceMaterial.hpp:76]: Done, had to update documentation as well though.

File: Tests/UnitTests/Core/Material/CMakeLists.txt (hunk 3 above)
Comment (asalzburger) [reply Tests/UnitTests/Core/Material/CMakeLists.txt]: Fixed.

**Comment by asalzburger** [comment]
Hey, I think I have addressed everything now.

**Review by niermann999 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [bbd7a1170d62c0212a8e80f9128065e1eeab389c]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35137778) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6710361)!

**Pull request merged**